src/main/java/uk/ac/cam/cl/dtg/isaac/dao/PgQuizQuestionAttemptPersistenceManager.java

```diff
@@ -112,4 +115,38 @@ public void registerQuestionAttempt(Long quizAttemptId, QuestionValidationRespon
             throw new SegueDatabaseException("Unable to process json exception", e);
         }
     }
+
+    @Override
+    public Map<Long, Map<String, List<QuestionValidationResponse>>> getAllAnswersForQuizAssignment(Long quizAssignmentId) throws SegueDatabaseException {
+        PreparedStatement pst;
+        try (Connection conn = database.getDatabaseConnection()) {
+            pst = conn.prepareStatement("SELECT user_id, question_id, question_attempt FROM quiz_question_attempts" +
+                " INNER JOIN quiz_attempts ON (quiz_attempts.id = quiz_question_attempts.quiz_attempt_id) WHERE quiz_assignment_id = ? ORDER BY quiz_attempt_id, timestamp");
+
+            pst.setLong(1, quizAssignmentId);
+
+            ResultSet results = pst.executeQuery();
+
+            Map<Long, Map<String, List<QuestionValidationResponse>>> resultsMap = Maps.newHashMap();
+            while (results.next()) {
+                Long userId = results.getLong("user_id");
+                String questionId = results.getString("question_id");
+
+                Map<String, List<QuestionValidationResponse>> userAttemptsMap = resultsMap.compute(userId, (ignoreKey, ignoreValue) -> Maps.newHashMap());
+
+                List<QuestionValidationResponse> questionAttempts = userAttemptsMap.computeIfAbsent(questionId, (ignore) -> Lists.newArrayList());
+
+                QuestionValidationResponse questionAttempt = objectMapper.readValue(
+                    results.getString("question_attempt"), QuestionValidationResponse.class);
+
+                questionAttempts.add(questionAttempt);
+            }
+
+            return resultsMap;
```

Comment:
We're never going to call it with a deleted assignment id, so I don't think it matters. Also, it would have to do potentially out of scope work to check if an assignment is deleted and return an error, which feels like a duplication?

Reply:
Sure! I am happy with that!